Ensure uniqueness of install prefixes in the database

haampie opened the pull request:
This pr handles install prefix collisions slightly better, and makes it such that the database can be used better as the 'source of truth' for what packages are installed.

With this PR if you install 2 packages `zlib~shared` and `zlib+shared` with projections that map these packages to the same directory, you get this:

```
$ spack -c "config:install_tree:root:$PWD" -c 'config:install_tree:projections:all:${PACKAGE}' install zlib~shared
...
==> zlib: Successfully installed zlib-1.2.11-k5l7ccyobzhdyxt3wpzzxgyy5va636sq
  Fetch: 0.16s.  Build: 0.58s.  Total: 0.74s.
[+] /tmp/tmp.EvXHRNxOxj/zlib

$ spack -c "config:install_tree:root:$PWD" -c 'config:install_tree:projections:all:${PACKAGE}' install zlib+shared
==> Error: <not overwriting install prefix from different spec>
```

Previously spack would delete the prefix directory of `zlib~shared` without warning, and you would end up with a missing install and a corrupted database where two zlib's would share the same install prefix.

This PR still handles partial installs (running spack install --keep-prefix xyz multiple times for a failing build where the build touches the install dir and you want to keep all that), since in that case the spec is not marked installed in the db, nor is the prefix path listed as available.

Finally it supersedes https://github.com/spack/spack/pull/24005, since with the install prefix collision check, we can discriminate between "the directory looks like a spack install of some package" and "the directory belongs to an installed spec".

Edit: would be nice to get some feedback to see if this is going in the right direction before adding better error handling and tests.

haampie committed:
Keep track of install prefixes in the db

This allows us to keep track of install prefix collisions.

haampie committed:
Guard against prefix collision

haampie committed:
We can trust the integrity of the db here to guard against both install prefix and sha1 collisions

haampie committed:
Ensure we track installed prefixes only

Uninstalled specs listed in the DB just do not occupy their install prefix directory, so other specs are free to use them

haampie committed:
Fix test

haampie commented:
FWIW, big environment installs have a nice speedup when most of the specs are installed already (yes i'm aware of the yaml -> json pr that would improve things too, but we'd skip parsing json as well.):

Before (checking yaml in install prefix):

https://asciinema.org/a/417714

After (querying the DB):

https://asciinema.org/a/417711

scheibelp commented on lib/spack/spack/database.py:
```diff
@@ -382,6 +382,11 @@ def __init__(self, root, db_dir=None, upstream_dbs=None,
+        # For every installed spec we keep track of its install prefix, so that
+        # we can answer the simple query whether a given path is already taken
+        # before installing a different spec.
+        self._prefixes = set()
```
Can you call this `self._installed_prefixes`?

scheibelp commented on lib/spack/spack/database.py:
```diff
@@ -784,6 +790,10 @@ def invalid_record(hash_key, error):
                 #       this?
                 data[hash_key] = InstallRecord.from_dict(spec, rec)
+
+                # Only mark a path as taken when the spec is installed
```
Based on suggested name change `installed_prefixes` I think this comment is unnecessary

scheibelp commented on lib/spack/spack/database.py:
```diff
@@ -1172,7 +1190,9 @@ def _decrement_ref_count(self, spec):
         rec.ref_count -= 1
 
         if rec.ref_count == 0 and not rec.installed:
+            # The install prefix has already been removed from self._prefixes
```
I think it may be best to add a unit test which ensures this and remove the comment.

scheibelp commented on lib/spack/spack/installer.py:
```diff
@@ -829,7 +836,7 @@ def _prepare_for_install(self, task):
             task.pkg.stage.destroy()
 
-        if not partial and self.layout.check_installed(task.pkg.spec) and (
+        if not partial and installed_in_db and (
```
I think you can remove `not partial` (based on https://github.com/spack/spack/pull/24005#discussion_r641861345)

scheibelp commented on lib/spack/spack/installer.py:
```diff
@@ -814,6 +814,13 @@ def _prepare_for_install(self, task):
         rec, installed_in_db = self._check_db(task.pkg.spec)
 
+        # Check for an install prefix collision
+        available = spack.store.db.is_occupied_install_prefix(task.pkg.spec.prefix)
```
(question, no change required) I think this could have also been fixed by placing a call to `check_installed` here. I'm curious how much this speeds things up for you (I imagine if you frequently reinstall large DAGs where many specs are already installed, this could consume some time for each already-installed spec).

haampie replied:
Yeah, I was trying to avoid the latter, see the asciinema links above and imagine the last spec is being dev'ed and reinstalled every now and then.

scheibelp reviewed (changes requested):
Overall I agree with the changes; if possible adding tests would be great (if you're short on time, then I can add them to this PR).

Based on checking into this, it seems that our collision checking logic is broken entirely (although it is only obvious when using a simplified projection). That was not clear to me from reading the PR description.

* check_installed in Database.add fails because it isn’t called until after the installation (and at that point the spec.yaml was overwritten)
* DirectoryLayout.create_install_directory still calls check_installed, but this appears to fail because we delete the prefix before doing the installation

I can’t immediately find any tests that ensure our collision detection works (but if there are tests, then they aren’t sufficient anyway, since I can confirm by going through `spack install` that this happens). Tests that would be good to add:

* Install different specs to the same projected prefix (i.e. your example)
* Install, uninstall, reinstall (make sure that there isn’t an erroneous collision - this would resolve my request in _decrement_ref_count)

haampie committed:
Rename to self._installed_prefixes and drop comments

haampie committed:
Simplify the logic

haampie committed:
Test install prefix collisions

We can have install prefix collisions even when there are no dag hash
collisions, since the user can specify a poor projection.

haampie commented:
1. Test: "Install, uninstall, reinstall" was already covered by `lib/spack/spack/test/cmd/uninstall.py`.
2. Test: "Install different specs to the same projected prefix" done
3. "I think you can remove not partial" done

haampie committed:
Make flake8 happy

haampie commented:
Can't reproduce these failures:

```
=========================== short test summary info ============================
FAIL lib/spack/spack/test/bindist.py::test_default_rpaths_create_install_default_layout
ERROR lib/spack/spack/test/bindist.py::test_default_rpaths_create_install_default_layout
```

Let me just close and reopen to see if it's spurious.

scheibelp commented:
Thanks! I realized while forming the commit message for this, that it does not detect prefix collisions when two different specs have the same hash. This is much rarer (in fact not anticipated to ever happen) so I have merged this as-is and plan to address that later.
